### ISSUE TITLE
Extend ParamUty to allow levels, lower, and upper

### ISSUE THREAD

**Issue by RaphaelS1** [mlr-org/paradox#251]
When I first used `paradox` and in particular `ParamUty` for adding untyped parameters, I had originally thought the point of this parameter was to allow any combination of `lower`, `upper`, `levels`, etc.

Of course in some cases this doesn't make sense because `ParamUty$new(lower = 1, upper = c(2,2))` would be meaningless however `ParamUty$new(levels = c(1,2,"a"))` is not.

Say I want to tune the `s` parameter in `predict.cv.glmnet`. In the `param_set` this is given as:
```R
ParamDbl$new(id = "s", lower = 0, upper = 1,
special_vals = list("lambda.1se", "lambda.min"), default = "lambda.1se",
tags = "predict")
```

It is not obvious to me how to tune this with the current available options, whereas if `ParamUty` could be constructed with levels then I would add to the autotuner something like:

```R
ParamUty$new("s", levels = c(0, 0.5, 1, "lambda.min"))
```

Which allows multiple input types and is intuitive. As a simpler example take tuning the number of trees in a random forest:

```R
ParamUty$new("ntrees", levels = c(100, 250, 750))
```

Two other options are:

1. Set the value in `ParamUty` not the levels, e.g.
```R
x = ParamDbl$new("ntrees")
x$value = c(100, 250, 750)
```
But I suspect this is more work as it then changes how `mlr3tuning` works internally

2. Allow `levels` in `ParamInt` and `ParamDbl`, e.g.

```R
ParamInt$new("ntrees", levels = c(100, 250, 750))
```

To be honest, to me the final option is the most preferable and it was the first thing I tried before realising it wasn't possible.

**Comment by jakob-r** [comment]
First some specific remarks to your examples:
- `c(0, 0.5, 1, "lambda.min")` should not be written like that because it converts numerics to characters.
- `ParamUty$new("s", levels = c(0, 0.5, 1, "lambda.min"))` does not support `levels` and it was not supposed to. Instead it was intended to use `ParamFct` in combination with a trafo to convert character values to whatever type you need them. I agree that this is cumbersome and we can have some nicer solution that automatically does this conversion. The tuner itself should however only see the factor values because it cannot now how to handle random object types.
- `ParamInt$new("ntrees", levels = c(100, 250, 750))` again this would either be a factor parameter because you this notation indirectly suggests that for the tuner the distance between 100 and 250 is the same as between 250 and 750. Again, `ParamFct` fails to handle the desired data type directly as it only allows characters so a trafo would be necessary. Better in this case would be to make a `ParamInt` from `1` to `3` and transform 1 to 100, 2 to 250 and 3 to 750. This directly shows the tuner what distance you suggest. Also this would totally change the behavior of `ParamInt` depending on how it is constructed which is a bad design choice IMHO.

Summing up: How to best define a search space also depends on the capabilities of the tuner. You can see PramSpaces from two perspectives: The tuning space for the tuner, and the domain for the learner. The tuning space should always be "simple" and have kind of meaningful distances between the parameters for most optimization algorithms.

tl,dr: We need easier ways to construct discrete numeric params.

**Comment by jakob-r** [comment]
I go with @berndbischl from the above referenced issue:

> it would probably best do
> 
> a) advice uses to do either this
> int [1,2,3] --trafo--> obj1, obj2, obj3 (if there is an order)
> fct ["a","b","c"] --trafo--> obj1, obj2, obj3 (if there is no order)
> 
> b) to have some sugar for the trafos, like above. this is already discussed here
> #245

**Comment by berndbischl** [comment]
hi @RaphaelS1 

1) important: Param objects are used for 2 purposes: describing, documenting and arg-asserting in leaners (*). and describing regions-of-interest for tuning (**). i guess you are talking about the later.

2) ParamUty was only meant for (*) and not (**). It is completely "unstructured" and hence cannot help to describe what should happen in tuning. Its mainly a "last-resort" for complex params, where we cannot express the type easily, but still register the param with the learner.

3) You seem to discretize a lot and mainly think about grid search..... For many tuners thats not really the natural approach.

4) Have you looked at the post above where jakob posted my suggestion from the other issue? that should cover all your cases?

**Comment by RaphaelS1** [comment]
> You seem to discretize a lot and mainly think about grid search..... For many tuners thats not really the natural approach.

Agreed but there are some good reasons for doing this, either by using known heuristics for parameters or in the interest of time/efficiency.

> Have you looked at the post above where jakob posted my suggestion from the other issue? that should cover all your cases?

Yes, I opened this issue as suggested as from a first-time-user perspective, it appeared as if `ParamUty` should be able to handle this case and the distinction between (*) and (**) is not immediately obvious as it seems strange to reconstruct a parameter set.


Not to go on about this repeatedly but I think this is a very good use-case for `set6`. And whilst we shouldn't discuss this in this issue I'm putting this here now to return to it later when we discuss the combination of the two. But by example say we want to tune the number of trees (integer), node size (integer) and splitting rule (factor) in `rfsrc`:
```R
ParamSet$new(
  ntree = Set$new(100, 200)
  nodesize = Interval$new(1, 5, "integer")
  splitrule = Set$new("mse", "quantile.regr")
)

Which is instantly clean and interpretable. But to be returned to in later discussions.

**Comment by mb706** [comment]
I suggest closing this issue.

**Comment by RaphaelS1** [comment]
I forgot it was open, happy to close but I think we should discuss this as a motivating example in our paradox/distr6 meeting.
